File: src/main/java/com/jackmeng/stl/stl_OutputCache.java
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package com.jackmeng.stl;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class stl_OutputCache< I, O >
{

  private final Function< I, O > function;
  private final Map< I, O > cache = new HashMap<>();

  public stl_OutputCache(Function< I, O > function)
  {
    this.function = function;
  }

  public O get(I input)
  {
    O output = cache.get(input);
    if (output == null)
    {
      output = function.apply(input);
      cache.put(input, output);
    }
    return output;
  }

  public void clearCache()
  {
    cache.clear();
  }
}

File: src/main/java/com/jackmeng/stl/stl_OutputCache.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

import java.util.HashMap
import java.util.function.Function

class stl_OutputCache<I , O>(private val function:Function<I , O>)
{
	private val cache:MutableMap<I , O?> =HashMap()
	operator fun get(input:I):O?
	{
		var output=cache[input]
		if (output==null)
		{
			output=function.apply(input)
			cache[input]=output
		}
		return output
	}
	
	fun clearCache()
	{
		cache.clear()
	}
}